File: common/schemas/shadow.py
"""Shadow Agent Detection schemas — analytics on denied gateway requests."""

import datetime

from pydantic import BaseModel

# ── Summary & List ──────────────────────────────────────────────────


class TopEndpointHit(BaseModel):
    """Endpoint probed by a shadow agent."""

    endpoint: str
    method: str
    count: int


class ShadowEvent(BaseModel):
    """A single denied request from a shadow agent."""

    id: int
    endpoint: str
    method: str
    deny_reason: str
    request_metadata: dict
    created_at: datetime.datetime | None


class ShadowAgentSummary(BaseModel):
    """Detected shadow agent in the list view."""

    agent_id: str
    deny_reason: str
    hit_count: int
    first_seen: datetime.datetime
    last_seen: datetime.datetime
    top_endpoints: list[str]
    is_blocked: bool = False
    is_dismissed: bool = False


class ShadowAgentListResponse(BaseModel):
    """Paginated list of detected shadow agents."""

    items: list[ShadowAgentSummary]
    total: int
    total_hits: int
    limit: int
    offset: int


# ── Detail ──────────────────────────────────────────────────────────


class ShadowAgentDetail(BaseModel):
    """Full detail view for a single shadow agent."""

    agent_id: str
    deny_reason: str
    hit_count: int
    first_seen: datetime.datetime
    last_seen: datetime.datetime
    top_endpoints: list[TopEndpointHit]
    recent_events: list[ShadowEvent]
    is_blocked: bool = False
    blocked_at: datetime.datetime | None = None
    is_dismissed: bool = False


# ── Stats ───────────────────────────────────────────────────────────


class ShadowAgentStats(BaseModel):
    """Summary counts for stats cards."""

    total_shadow_agents: int
    total_shadow_hits: int
    agents_not_found: int
    agents_inactive: int
    agents_blocked: int = 0
    agents_dismissed: int = 0


# ── Action requests/responses ───────────────────────────────────────


class BlockAgentRequest(BaseModel):
    """Request body for blocking a shadow agent."""

    reason: str | None = None


class BlockAgentResponse(BaseModel):
    """Response after blocking a shadow agent."""

    agent_id: str
    blocked: bool
    blocked_at: datetime.datetime


class DismissResponse(BaseModel):
    """Response after dismissing a shadow agent."""

    agent_id: str
    dismissed: bool

File: dashboard/src/services/api/shadow.ts
/**
 * Shadow Agent Detection API client.
 */

import { apiFetch, toQueryString } from './client'

// ── Types ───────────────────────────────────────────────────────────

export interface ShadowAgentStats {
  total_shadow_agents: number
  total_shadow_hits: number
  agents_not_found: number
  agents_inactive: number
  agents_blocked: number
  agents_dismissed: number
}

export interface ShadowAgentSummary {
  agent_id: string
  deny_reason: string
  hit_count: number
  first_seen: string
  last_seen: string
  top_endpoints: string[]
  is_blocked: boolean
  is_dismissed: boolean
}

export interface ShadowAgentList {
  items: ShadowAgentSummary[]
  total: number
  total_hits: number
  limit: number
  offset: number
}

export interface TopEndpointHit {
  endpoint: string
  method: string
  count: number
}

export interface ShadowEvent {
  id: number
  endpoint: string
  method: string
  deny_reason: string
  request_metadata: Record<string, unknown>
  created_at: string | null
}

export interface ShadowAgentDetail {
  agent_id: string
  deny_reason: string
  hit_count: number
  first_seen: string
  last_seen: string
  top_endpoints: TopEndpointHit[]
  recent_events: ShadowEvent[]
  is_blocked: boolean
  blocked_at: string | null
  is_dismissed: boolean
}

export interface BlockAgentResponse {
  agent_id: string
  blocked: boolean
  blocked_at: string
}

export interface DismissResponse {
  agent_id: string
  dismissed: boolean
}

// ── API Functions ───────────────────────────────────────────────────

export function getShadowAgentStats(params?: {
  start_date?: string
  end_date?: string
}): Promise<ShadowAgentStats> {
  return apiFetch<ShadowAgentStats>(`/api/v1/shadow-agents/stats${toQueryString(params || {})}`)
}

export function getShadowAgents(params: {
  limit?: number
  offset?: number
  start_date?: string
  end_date?: string
  min_hits?: number
  deny_reason?: string
  include_dismissed?: boolean
}): Promise<ShadowAgentList> {
  return apiFetch<ShadowAgentList>(`/api/v1/shadow-agents${toQueryString(params)}`)
}

export function getShadowAgentDetail(
  agentId: string,
  params?: { start_date?: string; end_date?: string },
): Promise<ShadowAgentDetail> {
  return apiFetch<ShadowAgentDetail>(
    `/api/v1/shadow-agents/${agentId}${toQueryString(params || {})}`,
  )
}

export function blockShadowAgent(agentId: string, reason?: string): Promise<BlockAgentResponse> {
  return apiFetch<BlockAgentResponse>(`/api/v1/shadow-agents/${agentId}/block`, {
    method: 'POST',
    body: JSON.stringify(reason ? { reason } : {}),
  })
}

export function unblockShadowAgent(agentId: string): Promise<void> {
  return apiFetch<void>(`/api/v1/shadow-agents/${agentId}/block`, { method: 'DELETE' })
}

export function dismissShadowAgent(agentId: string): Promise<DismissResponse> {
  return apiFetch<DismissResponse>(`/api/v1/shadow-agents/${agentId}/dismiss`, { method: 'POST' })
}

export function undismissShadowAgent(agentId: string): Promise<void> {
  return apiFetch<void>(`/api/v1/shadow-agents/${agentId}/dismiss`, { method: 'DELETE' })
}
